app/models/user.rb

```diff
@@ -1,9 +1,24 @@
 class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable
+         :recoverable, :rememberable, :validatable,
+         :omniauthable, omniauth_providers: [:google_oauth2]
 
   belongs_to :account, optional: true
   has_many :feedbacks
 
   validates :name, allow_blank: true, length: { minimum: 2, maximum: 20 }
+
+  def self.from_omniauth(access_token)
+    data = access_token.info
+    user = User.find_by(email: data['email']) ||
+           User.create(
+             name: data['name'] || data['nickname'],
+             email: data['email'],
+             avatar_url: data['image'],
+             password: Devise.friendly_token[0, 20],
+             provider: access_token.provider
+           )
+    user.create_account(name: user.name, email: user.email)
```

Comment:
The class method "from_omniauth" is supposed to **fetch** user, finding it in DB. But now it also does several side effects. Can we improve it so that the code is more clear and readable?

Reply:
After I added the logic for sign in, it became a serious mistake to create a new user account. Thx!